### PR TITLE
Vladimir Shevtsov. 17-react-native

### PR DESCRIPTION
**Pull request by shevv920**

### CODE REVIEW

**Comment by mikitasolo** [comment]
image
wait what?
i assume you should create .gitignore in your folder and add .expo to the .gitignore

**Comment by SobakaSlava** [comment]
i assume you should add `.expo` to the `.gitignore`

**Comment by mikitasolo** [comment]
> image
> wait what?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/App.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import React, { useEffect } from 'react';
+import { useDispatch, Provider } from 'react-redux';
+import { NavigationContainer } from '@react-navigation/native';
+import { fetchComics } from './resources/comics.action';
+
+import AppNavigation from './navigation';
+import store from './resources/store';
+
+const  App = () => {
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/App.js:9]: why 2 spaces?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import fetchMarvel from '../fetchMarvel';
+
+export const FETCH_COMICS_START = 'FETCH_COMICS_START';
+export const FETCH_COMICS_SUCCESS = 'FETCH_COMICS_SUCCESS';
+export const FETCH_COMICS_ERROR = 'FETCH_COMICS_ERROR';
+export const ADD_CHARACTER_TO_FAVORITES = 'ADD_CHARACTER_TO_FAVORITES';
+export const REMOVE_FROM_FAVORITES = 'REMOVE_FROM_FAVORITES';
+
+const fetchComicsStart = () => ({ type: FETCH_COMICS_START });
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js]: for what?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+const fetchComicsError = (err) => ({ type: FETCH_COMICS_ERROR, payload: err });
+
+export const fetchComics = (url, params) => (dispatch) => {
+  dispatch(fetchComicsStart());
+  fetchMarvel(url, params)
+    .then((res) => dispatch(fetchComicsSuccess(res)))
+    .catch((err) => dispatch(fetchComicsError(err)));
+};
+
+export const addToFavorites = (obj) => ({ type: ADD_CHARACTER_TO_FAVORITES, payload: obj });
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js]: this 2 actions can be united into the one

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  FETCH_COMICS_START,
+  FETCH_COMICS_SUCCESS,
+  FETCH_COMICS_ERROR,
+  ADD_CHARACTER_TO_FAVORITES,
+  REMOVE_FROM_FAVORITES,
+} from './comics.action';
+
+const initState = {
+  favorites: [],
+  fetching: false,
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js]: i don't see `data` in initialState, please add

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+export const comics = (state = initState, action) => {
+  switch (action.type) {
+    case FETCH_COMICS_START:
+      return { ...state, fetching: true };
+    case FETCH_COMICS_SUCCESS:
+      return { ...state, data: action.payload, fetching: false };
+    case FETCH_COMICS_ERROR:
+      return { ...state, error: action.payload, fetching: false };
+    case ADD_CHARACTER_TO_FAVORITES:
+      return { ...state, favorites: uniqBy([...state.favorites, action.payload], 'id') };
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js]: no need to store favorites separately. it's better to use flags like `isFavorite`

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { uniqBy } from 'lodash';
+
+export const getData = (state) => state.comics.data;
+export const getError = (state) => state.comics.error;
+export const getFetching = (state) => state.comics.fetching;
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js]:
```suggestion
export const getIsFetching = (state) => state.comics.isFetching;
```

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+export const getChars = (state) => {
+  const data = state.comics.data;
+  const results = data?.data?.data?.results ?? [];
+
+  const chars = uniqBy(results, "id").map((obj) => (
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js]: why uniqBy? what is the situation when we have items with the same id in results?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { uniqBy } from 'lodash';
+
+export const getData = (state) => state.comics.data;
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.selector.js]:
```suggestion
export const getComics = (state) => state.comics.data;
```

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    .then((res) => dispatch(fetchComicsSuccess(characterId, res)))
+    .catch((err) => dispatch(fetchComicsError(err)));
+};
+
+const fetchCharactersStart = () => ({ type: FETCH_CHARACTERS_START });
+const fetchCharactersSuccess = (data) => ({ type: FETCH_CHARACTERS_SUCCESS, payload: data });
+const fetchCharactersError = (err) => ({ type: FETCH_CHARACTERS_ERROR, payload: err });
+
+
+export const fetchCharacters = () => (dispatch) => {
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.action.js]: why is this in `comics.actions.js`?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+export const comics = (state = initState, action) => {
+  switch (action.type) {
+    case FETCH_COMICS_START:
+      return { ...state, fetching: true };
+    case FETCH_COMICS_ERROR:
+      return { ...state, error: action.payload };
+    case FETCH_COMICS_SUCCESS: {
+      const char = state.characters.find((c) => c.id === payload.characterId);
+      char.comics = payload.data;
+      console.log(payload.data);
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js]: .

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+      const chars = [...state.characters.filter((c) => c.id !== payload.characterId), char];
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js]: the order of chars will change after every fetching of comics. it's not good.

File: 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    case FETCH_COMICS_ERROR:
+      return { ...state, error: action.payload, fetching: false };
+    case ADD_CHARACTER_TO_FAVORITES:
+      return { ...state, favorites: uniqBy([...state.favorites, action.payload], 'id') };
+    case REMOVE_FROM_FAVORITES:
+      return { ...state, favorites: state.favorites.filter((c) => c.id !== action.payload) };
+    case FETCH_CHARACTERS_SUCCESS:
+      return {
+        ...state,
+        characters: uniqBy(action.payload, "id").map((obj) => (
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/resources/comics.reducer.js]: what is the situation when we have items with the same id?

File: 17-react-native/classworks/Shevtsov_Vladimir/task/screens/Details/Details.screen.js [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+import { fetchComics } from '../../resources/comics.action';
+import { getCharacter } from '../../resources/comics.selector';
+
+function DetailsScreen() {
+  const { params } = useRoute();
+  const { characterId } = params;
+  const favorites = useSelector(getFavorites);
+  const dispatch = useDispatch();
+  const character = useSelector(getCharacter(characterId));
+  const isFavorite = favorites.find((char) => char.id === character.id);
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/screens/Details/Details.screen.js]: this should be a selector

File: 17-react-native/classworks/Shevtsov_Vladimir/task/screens/Favourite/Favourite.screen.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { Text, View, Image } from 'react-native';
+import { useSelector, useDispatch } from 'react-redux';
+import { getFavorites, getFetching } from '../../resources/comics.selector';
+import { SafeAreaView } from 'react-native-safe-area-context';
+import { FlatList, TouchableHighlight } from 'react-native-gesture-handler';
+
+import styles from './Favourite.styles';
+import { removeFromFavorites } from '../../resources/comics.action';
```
Comment (mikitasolo) [inline 17-react-native/classworks/Shevtsov_Vladimir/task/screens/Favourite/Favourite.screen.js]: the file should be named `comics.actions`

**Review by mikitasolo — CHANGES_REQUESTED** [review]

**Review by mikitasolo — APPROVED** [review]

**Pull request merged**